torch/csrc/jit/graph_executor.cpp

```diff
@@ -0,0 +1,396 @@
+      // Why do I have to care here about output_nr? I just want to say
+      // grad_fn->setOutputTo(i, v.input_port());
+      grad_fn->next_functions.emplace_back(v.grad_fn() ? v.grad_fn() : v.grad_accumulator(), v.output_nr());
+    }
+    captureInputs(*grad_fn, inputs);
+
+    unwrapVariables(inputs);
+    tensor_list outputs;
+    InterpreterState(f).runOneStage(std::move(inputs), outputs);
+    wrapTensors(outputs);
+    // hookup the gradients for the output tensors that require gradients
+    // to the inputs to our gradient function df
+    // TODO - XXX - if any output is the same tensor multiple times, views have to be
+    // setup here. We need to refactor autograd until it is safe for
+    // tensors to be constructed without all the viewing infrastructure.
+    // this is currently intentionally not done here so we can get an idea of our
+    // perf before introducing overhead for correctness
+    for(auto idx : grad.df_input_vjps) {
+      autograd::Variable o(outputs[idx]);
+      auto impl = o.get();
+      // Note: we have to set this up in place, or we have to
+      // throw away and reallocate variables that were already created in
+      // wrapTensors. We should add an API for this, and more generally
+      // we need to clean up the fields of Variable.
+      impl->_grad_fn = grad_fn;
+      impl->output_nr = grad_fn->num_inputs++;
+      impl->_requires_grad = true;
+    }
+    captureOutputs(*grad_fn, outputs);
+    // drop the temporary outputs so that we return the same number of
+    // outputs as if we were not also calculating gradient
+    outputs.erase(outputs.begin() + grad.f_real_outputs, outputs.end());
+    return outputs;
+  }
+  Code f;
+  // description of gradient as a graph
+  Gradient grad; // if(grad) is false when this is unused
+  // executor for df, including code caches
+  GraphExecutor grad_executor;
+};
+
+} // anonymous namespace
+
+// a Graph can be created via tracing, or via a language-based frontend
+// GraphExecutor runs it. It can run the same graph on many different sizes
+// and different requires_grad states, and handles specializations for each situation.
+// GraphExecutor is completely unaware of tracing or module parameters to keep the
+// tracing concerns separated.
+struct GraphExecutorImpl {
+
+  GraphExecutorImpl(std::shared_ptr<Graph> graph, bool optimize, bool symbolically_differentiable)
+  : graph(std::move(graph))
+  , optimize(optimize)
+  , symbolically_differentiable(symbolically_differentiable) {}
+  GraphExecutorImpl(std::shared_ptr<Graph> graph, bool optimize)
+  : graph(std::move(graph))
+  , optimize(optimize)
+  , symbolically_differentiable(isDifferentiable(*this->graph)) {}
+
+  // entry point where execution begins
+  variable_tensor_list run(variable_tensor_list inputs) {
+    // this is the fallback pathway, when we cannot differentiate
+    if(!optimize || (!symbolically_differentiable && gradientIsPossible(inputs))) {
```

Comment:
do we really want to have `!optimize` here? I'd say that we should still run the graph through our symbolic pathways, we just shouldn't run any of our optimization passes on it.

Reply:
The intent of !optimize is to just run the graph without doing _anything_ to ensure that graph does what you think it does. It is effectively `-0`, and tries to do as little as possible so it doesn't go through pathways with potential bugs.